tls/s2n_resume.c

```diff
@@ -228,27 +228,53 @@ static int s2n_client_serialize_resumption_state(struct s2n_connection *conn, st
     return 0;
 }
 
-static S2N_RESULT s2n_tls12_client_deserialize_session_state(struct s2n_connection *conn, struct s2n_stuffer *from)
+static S2N_RESULT s2n_tls12_client_deserialize_session_state(struct s2n_connection *conn,
+        struct s2n_blob *ticket, struct s2n_stuffer *from)
 {
     RESULT_ENSURE_REF(conn);
     RESULT_ENSURE_REF(from);
 
-    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &conn->resume_protocol_version));
+    /* Operate on a copy of the connection to avoid mutating the connection on
+     * failure. We have tests in s2n_resume_test.c that prove this level of copy
+     * is sufficient.
+     */
+    struct s2n_crypto_parameters *secure = conn->secure;
+    RESULT_ENSURE_REF(secure);
+    struct s2n_connection resume_conn = *conn;
+    struct s2n_crypto_parameters resume_secure = *secure;
+    resume_conn.secure = &resume_secure;
+
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &resume_conn.resume_protocol_version));
 
     uint8_t *cipher_suite_wire = s2n_stuffer_raw_read(from, S2N_TLS_CIPHER_SUITE_LEN);
     RESULT_ENSURE_REF(cipher_suite_wire);
-    RESULT_GUARD_POSIX(s2n_set_cipher_as_client(conn, cipher_suite_wire));
+    RESULT_GUARD_POSIX(s2n_set_cipher_as_client(&resume_conn, cipher_suite_wire));
 
     uint64_t then = 0;
     RESULT_GUARD_POSIX(s2n_stuffer_read_uint64(from, &then));
 
-    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(from, conn->secrets.version.tls12.master_secret, S2N_TLS_SECRET_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(from, resume_conn.secrets.version.tls12.master_secret,
+            S2N_TLS_SECRET_LEN));
 
     if (s2n_stuffer_data_available(from)) {
         uint8_t ems_negotiated = 0;
         RESULT_GUARD_POSIX(s2n_stuffer_read_uint8(from, &ems_negotiated));
-        conn->ems_negotiated = ems_negotiated;
+        resume_conn.ems_negotiated = ems_negotiated;
+    }
+
+    DEFER_CLEANUP(struct s2n_blob client_ticket = { 0 }, s2n_free);
+    if (ticket) {
+        RESULT_GUARD_POSIX(s2n_dup(ticket, &client_ticket));
```

Comment:
Can you tell me why you need this dup? I just want to check, is it because the ticket parameter no longer has memory alloced for it?

Reply:
Correct. We don't allocate any memory for it until we know we're going to keep it.

I could still allocate memory for it up in s2n_client_deserialize_with_session_ticket like before. However, that complicates the lifetime of that memory. We'd either still need to allocate new memory here and always free the memory allocated by s2n_client_deserialize_with_session_ticket, or know back up in s2n_client_deserialize_with_session_ticket whether or not it needs to free its memory (so whether the ticket was TLS1.2 or TLS1.3). That just makes the responsibilities of the methods all messy.